webui/react/src/utils/data.test.ts

```diff
@@ -59,28 +60,35 @@ const asyncFn = async (): Promise<boolean> => {
   }
 };
 
-const tests = [
-  { type: [ Type.AsyncFn, Type.Fn ], value: asyncFn },
-  { type: [ Type.SyncFn, Type.Fn ], value: syncFn },
-  { type: [ Type.SyncFn, Type.Fn ], value: voidFn },
-  { type: [ Type.Map, Type.Object ], value: new Map() },
-  { type: [ Type.Map, Type.Object ], value: new Map([ [ 'a', 'value1' ], [ 'b', 'value2' ] ]) },
-  { type: [ Type.Map, Type.Object ], value: new Map([ [ 'x', -1 ], [ 'y', 1.5 ] ]) },
-  { type: Type.Primitive, value: 'Jalapeño' },
-  { type: [ Type.Number, Type.Primitive ], value: -3.14159 },
-  { type: [ Type.Number, Type.Primitive ], value: 1.23e-8 },
-  { type: [ Type.Number, Type.Primitive ], value: 0 },
-  { type: Type.Primitive, value: null },
-  { type: Type.Primitive, value: undefined },
-  { type: Type.Object, value: {} },
-  { type: Type.Object, value: { 0: 1.5, a: undefined, [Symbol('b')]: null } },
-  { type: [ Type.Set, Type.Object ], value: new Set() },
-  { type: [ Type.Set, Type.Object ], value: new Set([ 'abc', 'def', 'ghi' ]) },
-  { type: [ Type.Set, Type.Object ], value: new Set([ -1.5, Number.MAX_VALUE, null, undefined ]) },
-];
 const object = { a: true, b: null, c: { x: { y: -1.2e10 }, z: undefined } };
 
 describe('data utility', () => {
+  describe('flattenObject', () => {
+    const tests = [
+      {
+        input: {
+          a: {
+            x: true,
+            y: -5.280,
+            z: { hello: 'world' },
+          },
+          b: [ 0, 1, 2 ],
```

Comment:
that's a really good question...

@rb-determined-ai I didn't consider the case @hamidzr brings up. Can the hp dict key itself contain periods? So something like this would get returned by the API for hp JSON:
```
{
  a: {
    x: 'foo',
    y: 'bar',
  },
  'a.x': 123,
  'a.y': 456,
}
```
Definitely seems like an edge case but checking for the possibility that maybe this is prevented. I know @liamcli worked on modifying the hp-importance endpoint to flatten the hp keys there as well.

Reply:
Yes it is possible for a top-level key to contain a `.`.  There was a point in a PR against master where we were going to flatten/unflatten PRs like this, but we decided to only operate on them in a structured way (where they are only stored as go structs) because it was basically a lossy transition.

Overall, I think it's fine to flatten keys (the lossy part) but it's never fine to unflatten keys.  I.e. if you want to flatten them into a list to render that's ok, and if users do something dumb like have a `b.x: 2` and a `b: {x: 3}`, and it doesn't render right in the webui, eh.  They shouldn't do dumb things.  As long as we don't crash in those cases, I think it's not the end of the world if they don't render quite right.

Side note, if any API Endpoints on the master are emitting pre-flattened keys, we should address that before calling the feature 'complete'.

Reply:
> if any API Endpoints on the master are emitting pre-flattened keys;

is this in cases other that where the api is returning the experiment config? otherwise did we need any clientside changes?

Reply:
> Side note, if any API Endpoints on the master are emitting pre-flattened keys, we should address that before calling the feature 'complete'.

@rb-determined-ai Doesn't this [PR](https://github.com/determined-ai/determined/pull/2508) flatten the HP importance endpoint to return pre-flattened keys?